### PR TITLE
[SPARK-29983][SQL] Add an independent config for an optional INTERVAL clause

### PR DESCRIPTION
**Pull request by maropu**
### What changes were proposed in this pull request?

This pr is to add a new config only for an optional INTERVAL clause. In the master, this feature is enabled when `spark.sql.ansi.enabled`=true. This pr proposes to split off the optional interval flag from `spark.sql.ansi.enabled`.

This comes from the @cloud-fan suggestion: https://github.com/apache/spark/pull/26584#discussion_r347939499

### Why are the changes needed?

Actually, an INTERVAL keyword is not optional in the ANSI/SQL standard.
```
<interval literal> ::=
  INTERVAL [ <sign> ] <interval string> <interval qualifier>

<interval string> ::=
  <quote> <unquoted interval string> <quote>
```

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

No.

### CODE REVIEW

**Comment by maropu** [comment]
cc: @cloud-fan @gengliangwang

**Comment by cloud-fan** [comment]
one correction: the SQL standard actually requires the INTERVAL prefix
```
<interval literal> ::=
  INTERVAL [ <sign> ] <interval string> <interval qualifier>

<interval string> ::=
  <quote> <unquoted interval string> <quote>
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -1787,6 +1787,13 @@ object SQLConf {
     val SQL_STANDARD, ISO_8601, MULTI_UNITS = Value
   }
 
+  val OPTIONAL_INTERVAL =
+    buildConf("spark.sql.optionalInterval")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: let's put it under the `spark.sql.parser` namespace

**Comment by maropu** [comment]
> one correction: the SQL standard actually requires the INTERVAL prefix
Oh! I did misunderstand that.... thanks for the correction, @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 1 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: ok

**Comment by maropu** [comment]
The PR descripion updated based on the suggsetion above.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParseDriver.scala [hunk 2]
```diff
@@ -101,6 +101,7 @@ abstract class AbstractSqlParser(conf: SQLConf) extends ParserInterface with Log
     lexer.addErrorListener(ParseErrorListener)
     lexer.legacy_setops_precedence_enbled = conf.setOpsPrecedenceEnforced
     lexer.SQL_standard_keyword_behavior = SQLStandardKeywordBehavior
+    lexer.optional_interval = conf.optionalInterval
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParseDriver.scala]: I think it should be `dialect == Spark && ansi == true && optionalInterval`

For pgsql diaelct, INTERVAL can't be optional. for non-ansi mode, year/month/day... are not reserved keywords and we can't ommit INTERVAL prefix anyway.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 3]
```diff
@@ -1787,6 +1787,13 @@ object SQLConf {
     val SQL_STANDARD, ISO_8601, MULTI_UNITS = Value
   }
 
+  val OPTIONAL_INTERVAL =
+    buildConf("spark.sql.parser.optionalInterval")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: maybe "optionalIntervalPrefix"?

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: ok

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParseDriver.scala (hunk 2 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParseDriver.scala]: ok

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala [hunk 4]
```diff
@@ -255,8 +255,8 @@ class SQLQueryTestSuite extends QueryTest with SharedSparkSession {
 
     // If `--IMPORT` found, load code from another test case file, then insert them
     // into the head in this test.
-    val importedTestCaseName = comments.filter(_.startsWith("--IMPORT ")).map(_.substring(9))
-    val importedCode = importedTestCaseName.flatMap { testCaseName =>
+    val importedTestCaseNames = comments.filter(_.startsWith("--IMPORT ")).map(_.substring(9))
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala]: This is not related to this pr though, I found the misleading name here...

File: sql/core/src/test/resources/sql-tests/inputs/ansi/interval.sql [hunk 5]
```diff
@@ -1,17 +1,6 @@
+--SET spark.sql.parser.optionalIntervalPrefix=false
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/ansi/interval.sql]: do we still need this file?

Comment (maropu) [reply sql/core/src/test/resources/sql-tests/inputs/ansi/interval.sql]: I just checked if we could turn off the optional prefix with ansi enabled. I personally think keeping this is better because the parser logic is a bit complicated for the combination about ansi + optionalIntervalPrefix...

**Comment by cloud-fan** [comment]
sorry to have this proposal at this stage: can we drop this feature? It looks only useful when we have mysql/hive dialect someday, but I don't think it's going to happen in the near future. Even pgsql dialect is far away from completion.

**Comment by maropu** [comment]
ok
